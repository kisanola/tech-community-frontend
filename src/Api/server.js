import axios from 'axios';

const token = localStorage.getItem('token') || null;
const server = axios.create({
  baseURL: process.env.REACT_APP_BACKEND_URL,
  headers: {
    Authorization: token,
    'Content-Type': 'application/json',
  },
});

server.interceptors.request.use(
  async config => {
    const token = await localStorage.getItem('token');
    if (token) config.headers['Authorization'] = token;
    return config;
  },
  error => {
    return Promise.reject(error);
  },
);

export default server;
