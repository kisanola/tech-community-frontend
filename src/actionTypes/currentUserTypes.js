export const SET_CURRENT_USER = 'SET_CURRENT_USER';
export const SET_IS_AUTH = 'SET_IS_AUTH';
