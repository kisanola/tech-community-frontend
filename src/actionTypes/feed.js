export const TOOGLE_LOADING = 'TOOGLE_LOADING';
export const SET_FEED = 'SET_FEED';
