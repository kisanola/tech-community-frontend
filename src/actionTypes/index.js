export const CHANGE_NAME = 'CHANGE_NAME';
export * from './currentUserTypes';
