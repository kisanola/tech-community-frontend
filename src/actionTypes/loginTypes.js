export const NO_ERROR = 'NO_ERROR';
export const LOGIN = 'LOGIN';
export const LOGIN_STARTED = 'LOGIN_STARTED';
export const LOGIN_SUCCESS = 'LOGIN_SUCCESS';
export const LOGIN_ERROR = 'LOGIN_ERROR';
export const LOGIN_INPUT_CHANGED = 'LOGIN_INPUT_CHANGED';
