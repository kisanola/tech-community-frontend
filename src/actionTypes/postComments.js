export const TOGGLE_LOADING = 'TOGGLE_LOADING';
export const SET_COMMENTS = 'SET_COMMENTS';
