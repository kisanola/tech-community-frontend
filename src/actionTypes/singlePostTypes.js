export const TOGGLE_LOADING = 'TOGGLE_LOADING';
export const FETCH_SUCCESS = 'FETCH_SUCCESS';
