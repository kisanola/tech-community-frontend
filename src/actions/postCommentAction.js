import server from '../Api/server';
import {
  POST_COMMENT_ERROR,
  POST_COMMENT_STARTED,
  POST_COMMENT_SUCCESS,
  RESTORE_COMMENT_TICK,
} from '../actionTypes/postCommentTypes';
import { getPostComments } from './postComments';

/**
 * Triggers an action
 * when the process starts
 * @return {object} action
 */
const postCommentStarted = () => ({ type: POST_COMMENT_STARTED });

/**
 * Triggers an action
 * when the process succeed
 * @param {object} post
 * @return {object} action
 */
const postCommentSuccess = (post) => ({
  type: POST_COMMENT_SUCCESS,
  payload: post,
});

/**
 * Triggers an action
 * when the process fails
 * @param {object} error
 * @return {object} action
 */
const postError = (error) => ({
  type: POST_COMMENT_ERROR,
  payload: error,
});

/**
 * Triggers an action
 * when the post has been succeessful
 * to restore the tick
 * @return {object} action
 */
const restoreTick = () => ({ type: RESTORE_COMMENT_TICK });

/**
 * post
 * @param {*} { username, email, password }
 * @return {object} response
 */
const postCommentAction = ({ slug, value: body }) => async (dispatch) => {
  dispatch(postCommentStarted());

  try {
    const payload = {
      body,
    };

    const res = await server.post(`/posts/${slug}/comments`, payload);

    const { post } = res.data;
    dispatch(postCommentSuccess(post));
    dispatch(getPostComments(slug));

    setTimeout(() => {
      dispatch(restoreTick());
    }, 3000);
    return true;
  } catch (e) {
    if (e.response) {
      dispatch(postError(e.response.data.message));
      return false;
    }
    dispatch(postError('Please check your internet connection'));
    return false;
  }
};

export default postCommentAction;
