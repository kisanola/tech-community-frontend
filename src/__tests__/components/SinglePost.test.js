import React from 'react';
import { shallow } from 'enzyme';
import axios from 'axios';
import moxios from 'moxios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import server from '../../Api/server';
import postCommentAction from '../../actions/postCommentAction';
import singlePostReducer from '../../reducers/singlePost';
import { FETCH_SUCCESS } from '../../actionTypes/singlePostTypes';
import {
  SinglePost,
  mapStateToProps,
  mapDispatchToProps,
} from '../../components/SinglePost/SinglePost';

const mockStore = configureMockStore([thunk]);

let wrapper;

const props = {
  post: {
    slug: 'slug',
    author: {},
    userType: 'userType',
    description: 'description',
    likesCount: 1,
    createdAt: 'createdAt',
    push: () => {},
  },
  relatedPosts: {
    isEmpty: false,
    feed: [],
  },
  onFetchSinglePost: jest.fn(),
  slug: 'slug',
  history: { push: () => {} },
};

describe('Post.jsx', () => {
  test('should render Post.jx', () => {
    wrapper = shallow(<SinglePost {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('State', () => {
    let mapped;
    let dispatch;
    beforeEach(() => {
      dispatch = jest.fn();
      mapped = mapDispatchToProps(dispatch);
    });

    test('should return `mapStateToProps`', () => {
      const expectedState = {
        post: {
          slug: 'slug',
        },
      };
      const state = mapStateToProps({
        singlePost: { ...expectedState },
      });
      expect(state).toEqual(expectedState);
    });

    it('Should call onGetFeed', () => {
      mapped.onFetchSinglePost('slug');
      expect(dispatch).toHaveBeenCalled();
    });
  });
});

describe('postCommentAction.js actions', () => {
  let store;
  beforeEach(() => {
    store = mockStore({
      events: {},
    });

    moxios.install(axios);
  });

  afterEach(() => {
    moxios.uninstall(axios);
  });

  it('should call the post action with success', () => {
    jest.spyOn(server, 'post').mockResolvedValue({
      data: {
        post: {
          description: 'description',
          image: 'image',
        },
      },
    });
    store.dispatch(postCommentAction('slug', 'body')).then((res) => {
      expect(res).toBe(true);
    });
  });

  it('should call the post action with error', () => {
    jest.spyOn(server, 'post').mockRejectedValue({
      response: {
        data: { message: 'message' },
      },
    });
    store.dispatch(postCommentAction('slug', 'body')).then((res) => {
      expect(res).toBe(true);
    });
  });

  it('should call the post action with error', () => {
    jest.spyOn(server, 'post').mockRejectedValue();
    store.dispatch(postCommentAction('slug', 'body')).then((res) => {
      expect(res).toBe(true);
    });
  });
});

describe('Post reducer', () => {
  it('should test the reducer with type FETCH_SUCCESS', () => {
    const state = singlePostReducer({}, { type: FETCH_SUCCESS, payload: { slug: 'slug' } });
    expect(state.post.slug).toBe('slug');
  });

  it('should test the reducer with type default', () => {
    const state = singlePostReducer({}, { type: 'default', payload: {} });
    expect(state).toEqual(expect.any(Object));
  });
});
