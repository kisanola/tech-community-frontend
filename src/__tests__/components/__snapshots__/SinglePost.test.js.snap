// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post.jsx should render Post.jx 1`] = `ShallowWrapper {}`;
